File: exercises/global_navigation/web-template/map.py
import numpy as np
import math
from math import pi as pi
import cv2

class Map:
	def __init__(self, pose3d):
		self.pose3d = pose3d
	
	def RTx(self, angle, tx, ty, tz):
		RT = np.matrix([[1, 0, 0, tx], [0, math.cos(angle), -math.sin(angle), ty], 
						[0, math.sin(angle), math.cos(angle), tz], [0, 0, 0, 1]])
		return RT
		
	def RTy(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), 0, math.sin(angle), tx], [0, 1, 0, ty],
						[-math.sin(angle), 0, math.cos(angle), tz], [0, 0, 0, 1]])
		return RT
		
	def RTz(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), -math.sin(angle), 0, tx], [math.sin(angle), math.cos(angle), 0, ty],
						[0, 0, 1, tz], [0, 0, 0, 1]])
		return RT
		
	def RTFormula(self):
		RTz = self.RTz(pi/2, 50, 70, 0)
		return RTz
		
	def getTaxiCoordinates(self):
		
		pose = self.pose3d.getPose3d()
                #print(pose)
		x = pose.x 
		y = pose.y 
		#print("x : {} , y : {}".format(x,y))
		scale_y = 0.6  ; offset_y = 153
		y = scale_y * y + offset_y
		
		scale_x = -0.3 ; offset_x = 77
		x = scale_x * x + offset_x
                if x > 77:
                   t = x - 77
                   x = 77 - t
                else:
                   t = 77 - x
                   x = t + 77
                   
		#print("x : {} , y : {}".format(x,y))
		return y, x 
        
        def getTaxiAngle(self):
		pose = self.pose3d.getPose3d()
		rt = pose.yaw 
                #print(rt)
		ty = math.cos(-rt) - math.sin(-rt)
		tx = math.sin(-rt) + math.cos(-rt)
                #print("tx : {} , ty : {}".format(tx,ty))
		return tx, ty
	# Function to reset
	def reset(self):
		# Nothing to do, service takes care!
		pass
	

File: exercises/marker_visual_loc/cpp_lib/src/Map.cpp
#include "Map.hpp"
#include <cmath>

Map::Map(std::function<Pose3d()> pose_getter, std::function<Pose3d()> noisy_pose_getter)
    : pose_getter_(pose_getter), noisy_pose_getter_(noisy_pose_getter)
{
}

cv::Mat Map::rtx(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        1, 0, 0, tx,
        0, std::cos(angle), -std::sin(angle), ty,
        0, std::sin(angle), std::cos(angle), tz,
        0, 0, 0, 1);
}

cv::Mat Map::rty(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), 0, std::sin(angle), tx,
        0, 1, 0, ty,
        -std::sin(angle), 0, std::cos(angle), tz,
        0, 0, 0, 1);
}

cv::Mat Map::rtz(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), -std::sin(angle), 0, tx,
        std::sin(angle), std::cos(angle), 0, ty,
        0, 0, 1, tz,
        0, 0, 0, 1);
}

cv::Mat Map::rt_vacuum()
{
    return rtz(CV_PI / 2.0, 50.0, 70.0, 0.0);
}

std::optional<std::tuple<double, double, double>> Map::get_robot_coordinates()
{
    Pose3d pose = pose_getter_();
    
    if (pose.timeStamp == 0.0) {
        return std::nullopt;
    }

    double scale_y = -85.0;
    double offset_y = -6.88;
    double y = scale_y * (offset_y - pose.y);

    double scale_x = 83.0;
    double offset_x = 8.0;
    double x = scale_x * (offset_x - pose.x);

    return std::make_tuple(x, y, pose.yaw);
}

std::optional<std::tuple<double, double, double>> Map::get_robot_coordinates_with_noise()
{
    Pose3d pose = noisy_pose_getter_();
    
    if (pose.timeStamp == 0.0) {
        return std::nullopt;
    }

    double scale_y = -85.0;
    double offset_y = -6.88;
    double y = scale_y * (offset_y - pose.y);

    double scale_x = 83.0;
    double offset_x = 8.0;
    double x = scale_x * (offset_x - pose.x);

    return std::make_tuple(x, y, pose.yaw);
}

void Map::reset()
{
}